pinning.go

```diff
@@ -21,47 +21,79 @@ import (
 	"golang.org/x/xerrors"
 )
 
-func (cm *ContentManager) pinStatus(cont uint) (*types.IpfsPinStatus, error) {
-	var content Content
-	if err := cm.DB.First(&content, "id = ?", cont).Error; err != nil {
+// Get the status for each of a list of content IDs - returns an identically
+// sized array of statuses corresponding to the passed in IDs
+func (cm *ContentManager) pinStatuses(contentIDs []uint) ([]*types.IpfsPinStatus, error) {
+	statuses := make([]*types.IpfsPinStatus, len(contentIDs))
+
+	// Grab the available contents from the database
+	var dbContents []Content
+	if err := cm.DB.Find(&dbContents, "id IN ?", contentIDs).Error; err != nil {
 		return nil, err
 	}
 
-	cm.pinLk.Lock()
-	po, ok := cm.pinJobs[cont]
-	cm.pinLk.Unlock()
-	if !ok {
-		var meta map[string]interface{}
-		if content.PinMeta != "" {
-			if err := json.Unmarshal([]byte(content.PinMeta), &meta); err != nil {
-				log.Warnf("content %d has invalid pinmeta: %s", cont, err)
+	// Then write in status results
+	for i, id := range contentIDs {
+		cm.pinLk.Lock()
+		operation, ok := cm.pinJobs[id]
+		cm.pinLk.Unlock()
+
+		// If the status already exists in memory, use that
+		if ok {
+			statuses[i] = operation.PinStatus()
+		} else {
+			// Otherwise, find the content result from the database query...
```

Comment:
We could just sort the list and do it that way, Will the caller be surprised if the response order is different than the input?

Reply:
what do you mean @whyrusleeping?